In [1]:
import fuzzymatcher
import pandas as pd
from pathlib import *

hospital_accounts = pd.read_csv('hospital_account_info.csv')
hospital_reimbursement = pd.read_csv('hospital_reimbursement.csv')

The business scenario is that we want to match up the hospital reibursement information data with hospital customer data. Unfortunately in this scenario we do not have a common ID to join on so we will see if we can use these pythin tools to merge the data together based on a combination of name and address information. 
We will be using fuzzymatcher in this case. For this using the anaconda command prompt I have installed pandas, pathlib and fuzzy_matcher.

command to install on the anaconda command prompt is 
pip install (progame name)

In [2]:
#Let take a look at the hostpital data
hospital_accounts.head()

Account_Num                      Facility Name                    Address  \
0        10605             SAGE MEMORIAL HOSPITAL  STATE ROUTE 264 SOUTH 191   
1        24250        WOODRIDGE BEHAVIORAL CENTER       600 NORTH 7TH STREET   
2        10341           DOUGLAS GARDENS HOSPITAL            5200 NE 2ND AVE   
3        81101  SUNCOAST BEHAVIORAL HEALTH CENTER             4480 51ST ST W   
4        39835           TREASURE VALLEY HOSPITAL   8800 WEST EMERALD STREET   

           City State  ZIP Code County Name    Phone Number  \
0        GANADO    AZ     86505      APACHE  (928) 755-4541   
1  WEST MEMPHIS    AR     72301  CRITTENDEN  (870) 394-4113   
2         MIAMI    FL     33137  MIAMI-DADE  (305) 751-8626   
3     BRADENTON    FL     34210     MANATEE  (941) 792-2222   
4         BOISE    ID     83704         ADA  (208) 373-5000   

               Hospital Type              Hospital Ownership  
0  Critical Access Hospitals  Voluntary non-profit - Private  
1                Psychiatric                     Proprietary  
2       Acute Care Hospitals  Voluntary non-profit - Private  
3                Psychiatric                     Proprietary  
4       Acute Care Hospitals                     Proprietary

In [3]:
#now check out the reimbursement data
hospital_reimbursement.head()

Provider_Num                     Provider Name     Provider Street Address  \
0        839987  SOUTHEAST ALABAMA MEDICAL CENTER      1108 ROSS CLARK CIRCLE   
1        519118     MARSHALL MEDICAL CENTER SOUTH  2505 U S HIGHWAY 431 NORTH   
2        733073    ELIZA COFFEE MEMORIAL HOSPITAL          205 MARENGO STREET   
3        201752          MIZELL MEMORIAL HOSPITAL               702 N MAIN ST   
4        678488                 ST VINCENT'S EAST  50 MEDICAL PARK EAST DRIVE   

  Provider City Provider State  Provider Zip Code  Total Discharges  \
0        DOTHAN             AL              36301               118   
1          BOAZ             AL              35957                43   
2      FLORENCE             AL              35631                73   
3           OPP             AL              36467                12   
4    BIRMINGHAM             AL              35235                74   

   Average Covered Charges  Average Total Payments  Average Medicare Payments  
0                 20855.61                 5026.19                    4115.52  
1                 13289.09                 5413.63                    4490.93  
2                 22261.60                 4922.18                    4021.79  
3                 10901.33                 5343.50                    4284.17  
4                 28117.95                 5947.12                    4819.53

Since the columns have differnt names, I am definign account and reimbursement dataframes.

In [2]:
account = ["Facility Name", "Address", "City", "State"]
reimbursement = ["Provider Name", "Provider Street Address", "Provider City",
    "Provider State"]

In [5]:
account


['Facility Name', 'Address', 'City', 'State']

In [6]:
reimbursement

['Provider Name', 'Provider Street Address', 'Provider City', 'Provider State']

lets now apply fuzzy matching on the account

In [3]:
matched_result = fuzzymatcher.fuzzy_left_join (hospital_accounts,
                                               hospital_reimbursement,
                                               account,
                                               reimbursement,
                                               left_id_col='Account_Num',
                                               right_id_col='Provider_Num')

The above statement runs for around 2 mins.


In [5]:
matched_result

best_match_score  __id_left  __id_right  Account_Num  \
0              -0.746613      10605      643595        10605   
234            -0.609873      24250      426767        24250   
446            -0.404473      10341      730311        10341   
675            -0.333020      81101      302173        81101   
934            -0.629261      39835      416060        39835   
...                  ...        ...         ...          ...   
570968          1.272056      92281      501763        92281   
570969         -0.354010      65248      435284        65248   
571191         -0.753956      92377      752043        92377   
571456          1.873506      71562      597639        71562   
571457         -0.837109      51586      417133        51586   

                            Facility Name                         Address  \
0                  SAGE MEMORIAL HOSPITAL       STATE ROUTE 264 SOUTH 191   
234           WOODRIDGE BEHAVIORAL CENTER            600 NORTH 7TH STREET   
446              DOUGLAS GARDENS HOSPITAL                 5200 NE 2ND AVE   
675     SUNCOAST BEHAVIORAL HEALTH CENTER                  4480 51ST ST W   
934              TREASURE VALLEY HOSPITAL        8800 WEST EMERALD STREET   
...                                   ...                             ...   
570968      JEWISH HOSPITAL - SHELBYVILLE              727 HOSPITAL DRIVE   
570969         DAYTON CHILDREN'S HOSPITAL             ONE CHILDRENS PLAZA   
571191         NORTH TEXAS STATE HOSPITAL                  6515 KEMP BLVD   
571456               GULF BREEZE HOSPITAL           1110 GULF BREEZE PKWY   
571457            I-70 COMMUNITY HOSPITAL  105 HOSPITAL DRIVE, BUILDING B   

                 City State  ZIP Code County Name  ... Provider_Num  \
0              GANADO    AZ     86505      APACHE  ...       643595   
234      WEST MEMPHIS    AR     72301  CRITTENDEN  ...       426767   
446             MIAMI    FL     33137  MIAMI-DADE  ...       730311   
675         BRADENTON    FL     34210     MANATEE  ...       302173   
934             BOISE    ID     83704         ADA  ...       416060   
...               ...   ...       ...         ...  ...          ...   
570968    SHELBYVILLE    KY     40065      SHELBY  ...       501763   
570969         DAYTON    OH     45404  MONTGOMERY  ...       435284   
571191  WICHITA FALLS    TX     76308     WICHITA  ...       752043   
571456    GULF BREEZE    FL     32561  SANTA ROSA  ...       597639   
571457  SWEET SPRINGS    MO     65351      SALINE  ...       417133   

                             Provider Name  Provider Street Address  \
0                  TYLER MEMORIAL HOSPITAL  5950 STATE ROUTE 6 WEST   
234                CRISP REGIONAL HOSPITAL     902 7TH STREET NORTH   
446              JACKSON MEMORIAL HOSPITAL         1611 NW 12TH AVE   
675                   BLAKE MEDICAL CENTER           2020 59TH ST W   
934      ST LUKE'S REGIONAL MEDICAL CENTER  190 EAST BANNOCK STREET   
...                                    ...                      ...   
570968       JEWISH HOSPITAL - SHELBYVILLE       727 HOSPITAL DRIVE   
570969               MIAMI VALLEY HOSPITAL       ONE WYOMING STREET   
571191  UNITED REGIONAL HEALTH CARE SYSTEM         1600 11TH STREET   
571456                GULF BREEZE HOSPITAL    1110 GULF BREEZE PKWY   
571457      EVANGELICAL COMMUNITY HOSPITAL       ONE HOSPITAL DRIVE   

        Provider City Provider State Provider Zip Code Total Discharges  \
0         TUNKHANNOCK             PA             18657               18   
234           CORDELE             GA             31015               18   
446             MIAMI             FL             33136              113   
675         BRADENTON             FL             34209              135   
934             BOISE             ID             83712               68   
...               ...            ...               ...              ...   
570968    SHELBYVILLE             KY             40065               14   
570969  

The matched_results Dataframe contains best_match_score which shows the quality of the link.
Lets arrange the column for in eradable format for the top 5 matches.

In [8]:
cols =[ "best_match_score", "Facility Name", "Provider Name", "Address", "Provider Street Address",
    "Provider City", "City", "Provider State", "State"]
matched_result[cols].sort_values(by=['best_match_score'], ascending=False).head(5)

best_match_score                                      Facility Name  \
78178           3.090931    RARITAN BAY MEDICAL CENTER PERTH AMBOY DIVISION   
533317          2.799072            ROBERT WOOD JOHNSON UNIVERSITY HOSPITAL   
78767           2.785132  AVERA MCKENNAN HOSPITAL & UNIVERSITY HEALTH CE...   
242802          2.778860  JOHN T MATHER MEMORIAL HOSPITAL  OF PORT JEFFE...   
447786          2.721425               MAYO CLINIC HEALTH SYSTEM - RED WING   

                                            Provider Name  \
78178     RARITAN BAY MEDICAL CENTER PERTH AMBOY DIVISION   
533317            ROBERT WOOD JOHNSON UNIVERSITY HOSPITAL   
78767   AVERA MCKENNAN HOSPITAL & UNIVERSITY HEALTH CE...   
242802  JOHN T MATHER MEMORIAL HOSPITAL  OF PORT JEFFE...   
447786              MAYO CLINIC HEALTH SYSTEM IN RED WING   

                                       Address  \
78178                    530 NEW BRUNSWICK AVE   
533317           ONE ROBERT WOOD JOHNSON PLACE   
78767   1325 S CLIFF AVE  POST OFFICE BOX 5045   
242802                   75 NORTH COUNTRY ROAD   
447786         701 HEWITT BOULEVARD, PO BOX 95   

                       Provider Street Address   Provider City  \
78178                    530 NEW BRUNSWICK AVE     PERTH AMBOY   
533317           ONE ROBERT WOOD JOHNSON PLACE   NEW BRUNSWICK   
78767   1325 S CLIFF AVE  POST OFFICE BOX 5045     SIOUX FALLS   
242802                   75 NORTH COUNTRY ROAD  PORT JEFFERSON   
447786         701 HEWITT BOULEVARD, PO BOX 95        RED WING   

                  City Provider State State  
78178      PERTH AMBOY             NJ    NJ  
533317   NEW BRUNSWICK             NJ    NJ  
78767      SIOUX FALLS             SD    SD  
242802  PORT JEFFERSON             NY    NY  
447786        RED WING             MN    MN

In the above case if we look at the first record the with Best_match_score of 3.09. You can noticed that facility name and provider name, address, city and state are matching. Lets now look at the reverse of the above join.

In [9]:
matched_result[cols].sort_values(by=['best_match_score'], ascending=True).head(5)

best_match_score                                      Facility Name  \
426704         -2.268231                      CENTRO MEDICO WILMA N VAZQUEZ   
83262          -2.124071  DOCTOR CENTER HOSPITAL SAN FERNANDO DE LA CARO...   
42627          -2.106746     HOSPITAL ONCOLOGICO DR ISAAC GONZALEZ MARTINEZ   
450671         -2.050888           CENTRO DE SALUD CONDUCTUAL MENONITA-CIMA   
476557         -1.996508             ADMIN DE SERVICIOS MEDICOS  PUERTO RIC   

                           Provider Name  \
426704       BAPTIST MEDICAL CENTER EAST   
83262   OVERLAKE HOSPITAL MEDICAL CENTER   
42627             SCRIPPS MERCY HOSPITAL   
450671   MILFORD REGIONAL MEDICAL CENTER   
476557              MAINE MEDICAL CENTER   

                                                  Address  \
426704      CARR. 2 KM 39.5    ROAD NUMBER 2 BO ALGARROBO   
83262   EDIF JESUS T PINEIRO AVE FERNANDEZ JUNCOS BO P...   
42627   BO. MONACILLOS CARR 22 CENTRO MEDICO DE PUERTO...   
450671  CARR ESTATAL 14 INTERIOR SARGENTO GERARDO SANT...   
476557                           BO MONACILLO CARR NUM 22   

       Provider Street Address Provider City       City Provider State State  
426704         400 TAYLOR ROAD    MONTGOMERY  VEGA BAJA             AL    PR  
83262        1035-116TH AVE NE      BELLEVUE   CAROLINA             WA    PR  
42627             4077 5TH AVE     SAN DIEGO   SAN JUAN             CA    PR  
450671      14 PROSPECT STREET       MILFORD   AIBONITO             MA    PR  
476557          22 BRAMHALL ST      PORTLAND   SAN JUAN             ME    PR

Now similarly if you look at the value of best_match_score -2.268 you would notice that the facility name, address,city and state are not matching at all this will be a poor match or complete mismatch.

Now lets look at the extreme cases, lets's take a look at some of the matches that might be a little more challenging by looking at scores < 80:

In [11]:
matched_result[cols].query("best_match_score <=.80").sort_values(by=['best_match_score'], ascending=False).head(5)

best_match_score                  Facility Name  \
518788          0.792471       METHODIST HOSPITAL SOUTH   
417434          0.791668  ADVENTIST HEALTH UKIAH VALLEY   
303413          0.787163                 MADISON HEALTH   
388755          0.776632      PENN HIGHLANDS CLEARFIELD   
493279          0.775573  MEMORIAL HOSPITAL AT GULFPORT   

                              Provider Name                Address  \
518788  SOUTH TEXAS REGIONAL MEDICAL CENTER       1905 HWY 97 EAST   
417434          UKIAH VALLEY MEDICAL CENTER     275 HOSPITAL DRIVE   
303413          MADISON COUNTY HOSPITAL INC  210 NORTH MAIN STREET   
388755                  CLEARFIELD HOSPITAL       809 TURNPIKE AVE   
493279        MEMORIAL HOSPITAL AT GULFPORT       4500 13TH STREET   

          Provider Street Address Provider City        City Provider State  \
518788           1905 HWY 97 EAST    JOURDANTON  JOURDANTON             TX   
417434         275 HOSPITAL DRIVE         UKIAH       UKIAH             CA   
303413      210 NORTH MAIN STREET        LONDON      LONDON             OH   
388755           809 TURNPIKE AVE    CLEARFIELD  CLEARFIELD             PA   
493279  4500 13TH ST-P O BOX 1810      GULFPORT    GULFPORT             MS   

       State  
518788    TX  
417434    CA  
303413    OH  
388755    PA  
493279    MS

Looking at the above example we can say that fuzzymatcher isa good tool for a medium size data sets. As you start to get to 10000 of rows, it will take a lot of time to compute, so plan accordingly. however the eas of user- especially when working with pandas makes it a great first place to start.